File: Sources/PDFPageBox.m
#import "PDFPageBox.h"
#import "PDFDocument-Extension.h"
#import "TeX2img-Swift.h"

@interface PDFPageBox ()
{
    PDFPage *pdfPage;
}
@end

@implementation PDFPageBox
- (instancetype)initWithPDFPage:(PDFPage*)page
{
    if (!(self = [super init])) {
        return nil;
    } else {
        pdfPage = page;
    }
    return self;
}

+ (instancetype)pageBoxWithPDFPage:(PDFPage*)page
{
    return [[PDFPageBox alloc] initWithPDFPage:page];
}

+ (instancetype)pageBoxWithFilePath:(NSString*)path page:(NSUInteger)page
{
    PDFDocument *doc = [PDFDocument documentWithFilePath:path];
    if (!doc) {
        return nil;
    }
    return [doc pageAtIndex:page-1].pageBox;
}

- (NSRect)mediaBoxRect
{
    CGRect rect = CGPDFPageGetBoxRect(pdfPage.pageRef, kCGPDFMediaBox);
    return *(NSRect*)&rect;
}

- (NSRect)cropBoxRect
{
    CGRect rect = CGPDFPageGetBoxRect(pdfPage.pageRef, kCGPDFCropBox);
    return *(NSRect*)&rect;
}

- (NSRect)bleedBoxRect
{
    CGRect rect = CGPDFPageGetBoxRect(pdfPage.pageRef, kCGPDFBleedBox);
    return *(NSRect*)&rect;
}

- (NSRect)trimBoxRect
{
    CGRect rect = CGPDFPageGetBoxRect(pdfPage.pageRef, kCGPDFTrimBox);
    return *(NSRect*)&rect;
}

- (NSRect)artBoxRect
{
    CGRect rect = CGPDFPageGetBoxRect(pdfPage.pageRef, kCGPDFArtBox);
    return *(NSRect*)&rect;
}

- (NSString*)bboxStringOfBox:(CGPDFBox)boxType
                       hires:(BOOL)hires
                   addHeader:(BOOL)addHeader
{
    CGPDFPageRef pageRef = pdfPage.pageRef;
    CGRect mediaBoxRect = CGPDFPageGetBoxRect(pageRef, kCGPDFMediaBox);
    CGRect rect = CGRectIntersection(CGPDFPageGetBoxRect(pageRef, boxType), mediaBoxRect); // MediaBox でクリップ

    // gs がデフォルトで -dUseMediaBox で呼ばれることに対応して，MediaBox に対する相対座標を返す
    rect.origin.x -= mediaBoxRect.origin.x;
    rect.origin.y -= mediaBoxRect.origin.y;
   
    NSString *result;
    
    // 回転情報の考慮
    NSInteger rotation = pdfPage.rotation;
    if (rotation == 90) {
        rect = CGRectMake(rect.origin.y,
                          mediaBoxRect.size.width - rect.origin.x - rect.size.width,
                          rect.size.height,
                          rect.size.width);
    }
    if (rotation == 180) {
        rect = CGRectMake(mediaBoxRect.size.width - rect.origin.x - rect.size.width,
                          mediaBoxRect.size.height - rect.origin.y - rect.size.height,
                          rect.size.width,
                          rect.size.height);
    }
    if (rotation == 270) {
        rect = CGRectMake(mediaBoxRect.size.height - rect.origin.y - rect.size.height,
                          rect.origin.x,
                          rect.size.height,
                          rect.size.width);
    }
    
    if (hires) {
       result = [NSString stringWithFormat:@"%@%f %f %f %f\n",
                 (addHeader ? @"%%HiResBoundingBox: " : @""),
                 rect.origin.x,
                 rect.origin.y,
                 rect.origin.x + rect.size.width,
                 rect.origin.y + rect.size.height
                 ];
    } else {
        result = [NSString stringWithFormat:@"%@%ld %ld %ld %ld\n",
                  (addHeader ? @"%%BoundingBox: " : @""),
                  (NSInteger)floor(rect.origin.x),
                  (NSInteger)floor(rect.origin.y),
                  (NSInteger)floor(rect.origin.x) + (NSInteger)ceil(rect.size.width),
                  (NSInteger)floor(rect.origin.y) + (NSInteger)ceil(rect.size.height)
                  ];
    }
    
    return result;
}
@end

File: Sources/PDFPageBox.swift
import Quartz

class PDFPageBox: NSObject {
    let pdfPage: PDFPage
    let cgPdfPage: CGPDFPage
    
    init(pdfPage: PDFPage) {
        self.pdfPage = pdfPage
        self.cgPdfPage = pdfPage.pageRef!
        super.init()
    }
    
    @objc convenience init?(filePath path: String, page: UInt) {
        guard page > 0 else { return nil }
        let index = page - 1
        guard let pdfPage = PDFDocument(filePath: path)?.page(at: index) else { return nil }
        self.init(pdfPage: pdfPage)
    }
    
    func boxRect(_ box: CGPDFBox) -> NSRect {
        return self.cgPdfPage.getBoxRect(box) as NSRect
    }
    
    var mediaBoxRect: NSRect { self.boxRect(.mediaBox) }
    var cropBoxRect: NSRect { self.boxRect(.cropBox) }
    var bleedBoxRect: NSRect { self.boxRect(.bleedBox) }
    var trimBoxRect: NSRect { self.boxRect(.trimBox) }
    var artBoxRect: NSRect { self.boxRect(.artBox) }
    
    @objc func bboxString(of box: CGPDFBox, hires: Bool, addHeader: Bool) -> String {
        let mediaBoxRect = self.mediaBoxRect
        var rect = self.boxRect(box).intersection(mediaBoxRect) // MediaBox でクリップ
        
        // gs がデフォルトで -dUseMediaBox で呼ばれることに対応して，MediaBox に対する相対座標を返す
        rect.origin.x -= mediaBoxRect.origin.x
        rect.origin.y -= mediaBoxRect.origin.y
        
        let result: String
        
        // 回転情報の考慮
        let rotation = self.pdfPage.rotation
        if rotation == 90 {
            rect = CGRect(x: rect.origin.y,
                          y: mediaBoxRect.size.width - rect.origin.x - rect.size.width,
                          width: rect.size.height,
                          height: rect.size.width)
        }
        if (rotation == 180) {
            rect = CGRect(x: mediaBoxRect.size.width - rect.origin.x - rect.size.width,
                          y: mediaBoxRect.size.height - rect.origin.y - rect.size.height,
                          width: rect.size.width,
                          height: rect.size.height)
        }
        if (rotation == 270) {
            rect = CGRect(x: mediaBoxRect.size.height - rect.origin.y - rect.size.height,
                          y: rect.origin.x,
                          width: rect.size.height,
                          height: rect.size.width)
        }
        
        if hires {
            result = String(format: "%@%f %f %f %f\n",
                            (addHeader ? "%%HiResBoundingBox: " : ""),
                            rect.origin.x,
                            rect.origin.y,
                            rect.origin.x + rect.size.width,
                            rect.origin.y + rect.size.height)
        } else {
            result = String(format: "%@%ld %ld %ld %ld\n",
                            (addHeader ? "%%BoundingBox: " : ""),
                            Int(floor(rect.origin.x)),
                            Int(floor(rect.origin.y)),
                            Int(floor(rect.origin.x)) + Int(ceil(rect.size.width)),
                            Int(floor(rect.origin.y)) + Int(ceil(rect.size.height)))
        }
        
        return result
    }
}
